lib/streamlit/runtime/app_session.py

```diff
@@ -369,13 +369,9 @@ def request_rerun(self, client_state: ClientState | None) -> None:
             rerun_data = RerunData()
 
         if self._scriptrunner is not None:
-            if (
-                bool(config.get_option("runner.fastReruns"))
-                and not rerun_data.fragment_id_queue
-            ):
-                # If fastReruns is enabled and this is *not* a rerun of a fragment,
-                # we don't send rerun requests to our existing ScriptRunner. Instead, we
-                # tell it to shut down. We'll then spin up a new ScriptRunner, below, to
+            if bool(config.get_option("runner.fastReruns")):
+                # If fastReruns is enabled we don't send rerun requests to our existing ScriptRunner.
+                # Instead, we tell it to shut down. We'll then spin up a new ScriptRunner, below, to
                 # handle the rerun immediately.
```

Comment:
I think the unintended consequence of doing this is that long-running fragments will be preempted if a user interacts with another fragment. This may cause us to lose elements that would have been rendered in the app by the first fragment (note this isn't an issue with full script runs because the lost element ends up being rewritten in the full script run anyway).

I think a full solution using this approach would have to change the script runner to ignore the `request_stop` message if it's currently running a fragment, but that feels pretty dangerous to me. The reason we took the current approach with a fragment queue is that allowing multiple script runners to run simultaneously exposes the user to potentially having to think about thread safety when using fragments, which opens up a whole can of worms that we probably don't want to have users to have to worry about.

Reply:
Ahh thanks for the context! 
That makes sense to me! I have just tried to move the coalescing from [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/scriptrunner/script_requests.py#L118-L145) up to the `app_session.request_rerun` at [this place](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/app_session.py#L371) and preserve the  `widget_states` and `fragment_queue` also for a new thread if its currently running in a fragment. This works fine with regards to preempting the fragment (see test script), but now the issue is that if the user interacts too fast with the fragment before the whole app is rendered (in the example app the first slider), the other widgets are not rendered at all 😟 

<details>
<summary>Example App to simulate long-running fragment</summary>

```python
import time

import random


import streamlit as st

from streamlit.logger import get_logger

_LOGGER = get_logger(__name__)


@st.experimental_fragment
def my_longrunning_fragment():
    res = st.slider("Slide me (long)!")
    time.sleep(5)
    st.write(res)


@st.experimental_fragment  # (run_every=0.02)
def my_fragment():
    # st.write(time.time())
    # st.slider("Slide me!", value=random.randint(0, 9))
    res = st.slider("Slide me!")
    st.write(res)
    _LOGGER.debug(f"Finished fragment code: {time.time()}")


my_longrunning_fragment()
my_fragment()

res = st.slider("Full-app slider")
st.write(res)
```

</details>

Reply:
An alternative I looked into is that we could add a new flag `force_dgstack_cleaning` or similar to `RerunData`, set it in st.rerun [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/commands/execution_control.py#L73-L75), which is our main way to get out of fragments (and therefore dialogs) and clean the `dg_stack` [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/scriptrunner/script_runner.py#L591) when this flag is set. However, also here we would need to investigate side-effects. But if we only set the flag in `st.rerun`, there should not be too many side-effects because it triggers a whole app rerun (at least for now).

Reply:
Hm, yeah. My gut feeling is that the safer fix would involve resetting `dg_stack` rather than trying to make changes to the fragment_queue or similar. Likely worth further discussion, though.

Reply:
@vdonato I have updated the PR to reset the dg_stack; I think this is the right solution!
I have also added unit tests to simulate the behavior and fix! What do you think?